src/core/proof.ml

```diff
@@ -4,103 +4,209 @@ open! Lplib
-      ; goal_type : term  (* Precomputed type for a suitable term.  *) }
+    type goal_typ =
+      { goal_meta : meta  (* Metavariable needing instantiation.          *)
+      ; goal_hyps : Env.t (* Precomputed scope for a suitable term.       *)
+      ; goal_type : term  (* Precomputed type for a suitable term.        *) }
 
-    let of_meta : meta -> t = fun m ->
+    type t =
+      | Typ of goal_typ (* The usual proof type goal. *)
+      | Unif of constr (* Two terms we'd like equal in ctxt. *)
+
+    let is_typ  = function Typ _ -> true  | Unif _ -> false
+    let is_unif = function Typ _ -> false | Unif _ -> true
+    let typ = function x -> (Typ x)
+    let unif = function x -> (Unif x)
+    let goal_typ_of = function
+      | Typ  gt -> gt
+      | _ -> failwith "Internal error : not a type goal"
+    let constr_of   = function
+      | Unif cs -> cs
+      | _ -> failwith "Internal error : not an unification goal"
+
+    let goal_typ_of_meta : meta -> goal_typ = fun m ->
       let (goal_hyps, goal_type) =
         Infer.destruct_prod m.meta_arity !(m.meta_type)
       in
       let goal_type = Eval.simplify (Env.to_ctxt goal_hyps) goal_type in
       {goal_meta = m; goal_hyps; goal_type}
 
-    let get_meta : t -> meta = fun g -> g.goal_meta
+    let get_meta : goal_typ -> meta = fun g -> g.goal_meta
 
-    let get_type : t -> Env.t * term = fun g -> (g.goal_hyps, g.goal_type)
+    let get_type : goal_typ -> Env.t * term = fun g ->
+      (g.goal_hyps, g.goal_type)
 
-    let simpl : t -> t = fun g ->
+    let simpl : goal_typ -> goal_typ = fun g ->
       {g with goal_type = Eval.snf (Env.to_ctxt g.goal_hyps) g.goal_type}
 
-    let compare : t -> t -> int = fun g1 g2 ->
+    let compare : goal_typ -> goal_typ -> int = fun g1 g2 ->
       Meta.compare g1.goal_meta g2.goal_meta
   end
 
 (** Representation of the proof state of a theorem. *)
 type proof_state =
   { proof_name     : Pos.strloc  (** Name of the theorem.                 *)
-  ; proof_term     : meta        (** Metavariable holding the proof term. *)
+  ; proof_term     : meta option (** Metavariable holding the proof term. *)
   ; proof_goals    : Goal.t list (** Open goals (focused goal is first).  *) }
 
 (** Short synonym for qualified use. *)
 type t = proof_state
 
-(** [init builtins name a] returns an initial proof state for a theorem  named
-    [name], which statement is represented by the type [a]. Builtin symbols of
-    [builtins] may be used by tactics, and have been declared. *)
-let init : Pos.strloc -> term -> t = fun name a ->
-  let proof_term = Meta.fresh ~name:name.elt a 0 in
-  let proof_goals = [Goal.of_meta proof_term] in
-  {proof_name = name; proof_term; proof_goals}
+(** [goal_of_meta m] returns a goal associated to the meta m *)
+let goal_of_meta : meta -> Goal.t = fun m ->
+  Goal.Typ (Goal.goal_typ_of_meta m)
+
+(** [goals_of_metas m] returns the goals associated to the metas in a *)
```

Comment:
typo: m -> a

Reply:
Perhaps it would be better to rename goals_of_metas into goals_of_term to avoid confusion with goal_of_meta.

Reply:
this function was unused and has been removed